oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/MemoryManager.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.index.indexer.document.flatfile;
+
+import java.util.Optional;
+
+/**
+ * Implementations of this interface can be used to get updates about remaining heap memory.
+ */
+public interface MemoryManager {
+    /**
+     * Indicates type of memory management this instance does.
+     */
+    enum Type {
+        /**
+         * Uses JMX memory management mbeans for tracking available memory.
+         */
+        JMX_BASED,
+        /**
+         * Maintains memory usage itself. Different users can share this instance and indicate the increase/decrease in
+         * their memory usage by calling {@link #changeMemoryUsedBy(long)}, and check if the total memory used by all
+         * users of this calls causes the available memory to become low using {@link #isMemoryLow()}
+         */
+        SELF_MANAGED
+    }
+
+    Type getType();
+
+    /**
+     * Register a client with this memory manager. All registered clients are informed when available memory level is low.
+     * If memory level is already low, client is not registered.
+     * NOTE - this method should only be used with {@link Type#JMX_BASED} instance types, otherwise the behaviour is undefined.
```

Comment:
I would just write that "otherwise, it has no effect". Or maybe reverse it and say, "for SELF_MANAGED, this method has no effect".

It is easier to understand I think. I don't see a reason for undefined behavior...

Reply:
I think it is very good to move memory management to a different class BTW.